In [1]:
############################ JAS'S CODE START #####################################################
import json
import re

with open('./full_course_list.json') as f:
    data = json.load(f)


In [2]:
def findall(p, s):
    '''Yields all the positions of
    the pattern p in the string s.'''
    var = s.find(p)
    while var != -1:
        yield var
        var = s.find(p, i+1)
        
def find_duplicates(major_key, string, index, relist):
    relist.append(major_key + " " + string[index:index+4])
    if string[index+4:index+5] == ',':
        find_duplicates(major_key,string,index+5,relist)
        
def helper(major_key, string, index):
    relist = []
    find_duplicates(major_key, string, index,relist)
    return relist



In [3]:
outside_major = ['Graduate/professional standing', 'Quantitative Reasoning (QR)','Capstone Certificate', 'Consent of instructor', 'Sophomore standing']
key_major_codes = ['COMP SCI', 'MATH', 'STAT']
data_for_mongo = []
db_change = []
for k in data:
    course_name = k['courseDesignation']
    lt = k['lastTaught']
    credit = k['creditRange']
    des = k['description']
    title = k['title']
    pre_req = k['enrollmentPrerequisites']
    # we are using a set so that we only add unique course numbers as the algo below may produce duplicates
    all_pre = set()
    for j in key_major_codes:
        #this finds all the major key codes in the string
        for i in findall(j, pre_req):

            #below are a bunch of different edge cases for the type of string that could follow the major
            if pre_req[i+len(j):i+len(j)+10] == '/MATH/STAT':
                pre = pre_req[i:i+len(j)] + " " + pre_req[i+len(j)+11:i+len(j)+14] 
                all_pre.update([pre])
                pre = pre_req[i+len(j)+1:i+len(j)+5] + " " + pre_req[i+len(j)+11:i+len(j)+14] 
                all_pre.update([pre])
                pre = pre_req[i+len(j)+6:i+len(j)+10] + " " + pre_req[i+len(j)+11:i+len(j)+14] 
                all_pre.update([pre])
            elif pre_req[i+len(j):i+len(j)+5] == '/MATH' or pre_req[i+len(j):i+len(j)+5] == '/STAT':
                pre = pre_req[i:i+len(j)] + " " + pre_req[i+len(j)+6:i+len(j)+9] 
                all_pre.update([pre])
                pre = pre_req[i+len(j)+1:i+len(j)+5] + " " + pre_req[i+len(j)+6:i+len(j)+9] 
                all_pre.update([pre])
            elif pre_req[i+len(j):i+len(j)+6] == '/E C E':  
                pre = pre_req[i:i+len(j)] + " " + pre_req[i+len(j)+7:i+len(j)+10] 
                all_pre.update([pre])
                pre = pre_req[i+len(j)+1:i+len(j)+6] + " " + pre_req[i+len(j)+7:i+len(j)+10] 
                all_pre.update([pre])
            elif pre_req[i+len(j)+5:i+len(j)+7] == 'or':
                try:
                    temp = int(pre_req[i+len(j)+8:i+len(j)+11])
                    all_pre.update([j + " " + str(temp) ])
                except:
                    continue
            elif pre_req[i+len(j)+5:i+len(j)+8] == 'and':
                try:
                    temp = int(pre_req[i+len(j)+9:i+len(j)+12])
                    all_pre.update([j + " " + str(temp)] )
                except:
                    continue
            elif pre_req[i+len(j)+5:i+len(j)+6] == ',':
                mult = helper(j,pre_req, i+len(j)+7)
                for b in mult:
                    all_pre.update([b])
            else:
                pre = pre_req[i:i+len(j)+4]
                all_pre.update([pre])
    # this checks for any requirments outside the courses
    for j in outside_major:
        if j in pre_req:
            all_pre.update([j])
    ############################### JAS'S CODE END ##########################################################
    
    
    ############################### MATT'S CODE START #######################################################
    # deleting the courses that should not be treated as prereqs from the all_pre list
    # (oppositely, these are the courses that couldn't have been taken previously to enroll in this course)
    if('Not open' in pre_req):
        # remembering the course which we should later update with regards to the current DB contents
        db_change += [course_name]
        ind = pre_req.find('Not open')
        excl = pre_req[ind:]
        courses = excl.split('COMP SCI')[1]
        nums = re.findall(r'\d+', courses)
        for num in nums:
            prereq_del = 'COMP SCI ' + str(num)
            if (prereq_del in all_pre):
                all_pre.remove(prereq_del)
    
    # same with the above if-statement but different phrasing in the json courses file
    if('does not' in pre_req):
        # remembering the course which we should later update with regards to the current DB contents
        db_change += [course_name]
        all_pre.remove('MATH 118')
    ############################### MATT'S CODE END #######################################################
    
    ############################### JAS'S CODE START ######################################################
    # this creates the json object we would want to store into mongoDB
    row = {"courseNumber":course_name,"info":{"courseName": title, "description":des, "credits":credit, "lastTaught":lt}, "prerequisites":list(all_pre)}
    data_for_mongo.append(row)
    ############################### JAS'S CODE END ########################################################
    

In [4]:
############################### MATT'S CODE START #######################################################
# visualizing the courses we should update to solve the prereqs bug (in which unneeded courses were treated as prereqs)
db_change = set(db_change)
db_change

{'COMP SCI 202',
 'COMP SCI 220',
 'COMP SCI 270',
 'COMP SCI 300',
 'COMP SCI 570'}

In [15]:
# Updating the courses DB to solve the bug
import pymongo

myclient = pymongo.MongoClient(
    "mongodb+srv://devUser:rIxwjrDdQUd5BKoT@cluster0.lwnmb.mongodb.net/madflow?retryWrites=true&w=majority&ssl=true&ssl_cert_reqs=CERT_NONE"
)
changed = []
for course_num in db_change:
    for course_row in data_for_mongo:
        if ((course_row['courseNumber'] == course_num) and (course_num not in changed)):
                changed += [course_num]
                myclient["madflow"]["courses"].insert_one(course_row)     
        

In [24]:
# Checking that there are no duplicates in Course collection of MongoDB
all_courses = []
cursor = myclient["madflow"]["courses"].find({})
for entry in cursor:
    all_courses += [entry['courseNumber']]
print('No course duplicates in MongoDB: ' + str(len(all_courses) == len(set(all_courses))))
############################### MATT'S CODE END #######################################################

No course duplicates in MongoDB: True


In [ ]:
# THE REST IS JAS'S CODE AS WELL
import pymongo

myclient = pymongo.MongoClient(
    "mongodb+srv://devUser:rIxwjrDdQUd5BKoT@cluster0.lwnmb.mongodb.net/madflow?retryWrites=true&w=majority&ssl=true&ssl_cert_reqs=CERT_NONE"
)

added = []
add_to_DB = []

for course in data_for_mongo:
    if course["courseNumber"] not in added:
        added.append(course["courseNumber"])
        add_to_DB.append(course)
add_to_DB


In [ ]:
# test = {'courseNumber': 'COMP SCI 200', 'info': {'courseName': 'Programming I', 'description': 'Learn the process of incrementally developing small (200-500 lines) programs along with the fundamental Computer Science topics. These topics include: problem abstraction and decomposition, the edit-compile-run cycle, using variables of primitive and more complex data types, conditional and loop-based flow control, basic testing and debugging techniques, how to define and call functions (methods), and IO processing techniques. Also teaches and reinforces good programming practices including the use of a consistent style, and meaningful documentation. Intended for students who have no prior programming experience. Enroll Info: None', 'credits': '3', 'lastTaught': '1222'}, 'prerequisites': ['Quantitative Reasoning (QR)','test']}
myclient["madflow"]["courses"].insert_many(add_to_DB)


(COMP SCI 367 or 400) and (COMP SCI 407, 536, 537, 545, 559, 564, 570, 679 or COMP SCI/E C E 552) or graduate/professional standing, or declared in the Capstone Certificate in Computer Sciences for Professionals

COMP SCI/MATH/STAT 475

COMP SCI/E C E 552

In [ ]:
def find_duplicates_majors(string, index, mjlist):
    string[]
    mjlist.append(major_key + " " + string[index:index+4])
    if string[index+4:index+5] == ',':
        find_duplicates(major_key,string,index+5,relist)
        
def helper(major_key, string, index):
    relist = []
    find_duplicates(major_key, string, index,relist)
    return relist